src/Membership.sol

```diff
@@ -0,0 +1,427 @@
+        /// @notice duration of the grace period
+        uint32 gracePeriod;
+        /// @notice the user message limit of each member
+        uint32 userMessageLimit;
+        /// @notice the index of the member in the set
+        uint32 index;
+        /// @notice address of the owner of this membership
+        address holder;
+        /// @notice token used to acquire this membership
+        address token;
+    }
+
+    /// @notice Emitted when a membership is erased due to having exceeded the grace period or the owner having chosen
+    /// to not extend it
+    /// @param idCommitment the idCommitment of the member
+    /// @param userMessageLimit the rate limit of this membership
+    /// @param index the index of the membership in the merkle tree
+    event MemberExpired(uint256 idCommitment, uint32 userMessageLimit, uint32 index);
+
+    /// @notice Emitted when a membership in grace period is extended
+    /// @param idCommitment the idCommitment of the member
+    /// @param userMessageLimit the rate limit of this membership
+    /// @param index the index of the membership in the merkle tree
+    /// @param newExpirationDate the new expiration date of this membership
+    event MemberExtended(uint256 idCommitment, uint32 userMessageLimit, uint32 index, uint256 newExpirationDate);
+
+    /// @dev contract initializer
+    /// @param _priceCalculator Address of an instance of IPriceCalculator
+    /// @param _maxTotalRateLimitPerEpoch Maximum total rate limit of all memberships in the tree
+    /// @param _minRateLimitPerMembership Minimum rate limit of one membership
+    /// @param _maxRateLimitPerMembership Maximum rate limit of one membership
+    /// @param _expirationTerm Membership expiration term
+    /// @param _gracePeriod Membership grace period
+    function __Membership_init(
+        address _priceCalculator,
+        uint32 _maxTotalRateLimitPerEpoch,
+        uint32 _minRateLimitPerMembership,
+        uint32 _maxRateLimitPerMembership,
+        uint32 _expirationTerm,
+        uint32 _gracePeriod
+    )
+        internal
+    {
+        priceCalculator = IPriceCalculator(_priceCalculator);
+        maxTotalRateLimitPerEpoch = _maxTotalRateLimitPerEpoch;
+        maxRateLimitPerMembership = _maxRateLimitPerMembership;
+        minRateLimitPerMembership = _minRateLimitPerMembership;
+        billingPeriod = _expirationTerm;
+        gracePeriod = _gracePeriod;
+    }
+
+    /// @notice Checks if a user message limit is valid. This does not take into account whether we the total
+    /// memberships have reached already the `maxTotalRateLimitPerEpoch`
+    /// @param userMessageLimit The user message limit
+    /// @return true if the user message limit is valid, false otherwise
+    function isValidUserMessageLimit(uint32 userMessageLimit) external view returns (bool) {
+        return userMessageLimit >= minRateLimitPerMembership && userMessageLimit <= maxRateLimitPerMembership;
```

Comment:
This does not take into account the expired memberships, although since it's an external function I could attempt to optionally calculate whether the rate limit of currently expired memberships would allow the userMessageLimit to be registered. WDYT?